File: game/tv_show/video_state.py
from audio_helper import AudioHelper
from phone_events import PhoneEvents
from state import State


class VideoState(State):
    def __init__(self, context, video, loop = False, audio_resource = None):
        super().__init__(context)
        self.video = video
        self.loop = loop
        self.has_looped = False
        self.audio_resource = audio_resource
        self.audio_instance_id = None

    def on_enter(self) -> None:
        super().on_enter()
        self.video.restart()
        if self.audio_resource:
            self.audio_instance_id = AudioHelper.play(self.audio_resource, self.context.audio_port, loops=-1)

    def process_events(self, phone_events: PhoneEvents):
        if self.loop and not self.video.active:
            self.video.restart()
            self.has_looped = True
            # Restart audio if it's not already playing
            if self.audio_resource and self.audio_instance_id is None:
                self.audio_instance_id = AudioHelper.play(self.audio_resource, self.context.audio_port, loops=-1)

    def render(self, screen):
        self.video.draw(screen, (0, 0))

    def on_exit(self) -> None:
        super().on_exit()
        self.video.stop()
        if self.audio_instance_id is not None:
            AudioHelper.stop(self.audio_instance_id, self.context.audio_port)
            self.audio_instance_id = None

    def video_ended(self):
        return not self.video.active

File: game-web/src/tvShow/videoState.ts
/** Port of game/tv_show/video_state.py */

import { State, type StateClass } from '../state';
import { AudioHelper } from '../audioHelper';
import type { VideoWrapper } from './tvShowResources';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class VideoState extends State {
  protected video: VideoWrapper;
  protected loop: boolean;
  protected hasLooped = false;

  private audioResource: string | null;
  private audioInstanceId: number | null = null;

  constructor(
    context: GameData,
    video: VideoWrapper,
    loop = false,
    audioResource: string | null = null,
  ) {
    super(context);
    this.video = video;
    this.loop = loop;
    this.audioResource = audioResource;
  }

  onEnter(): void {
    super.onEnter();
    this.video.restart();
    if (this.audioResource) {
      this.audioInstanceId = AudioHelper.play(this.audioResource);
    }
  }

  processEvents(_events: PhoneEvents): StateClass | null {
    if (this.loop && !this.video.active) {
      this.video.restart();
      this.hasLooped = true;
      if (this.audioResource && this.audioInstanceId === null) {
        this.audioInstanceId = AudioHelper.play(this.audioResource);
      }
    }
    return null;
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    // readyState < HAVE_CURRENT_DATA (2) means no frame is available yet
    if (this.video.element.readyState >= 2) {
      this.video.draw(ctx, 0, 0);
    }
  }

  onExit(): void {
    super.onExit();
    this.video.stop();
    if (this.audioInstanceId !== null) {
      AudioHelper.stop(this.audioInstanceId);
      this.audioInstanceId = null;
    }
  }

  protected videoEnded(): boolean {
    return !this.video.active;
  }
}
